src/cgame/rocket/rocket.cpp

```diff
@@ -525,99 +538,60 @@ Rml::String Rocket_QuakeToRML( const char *in, int parseFlags = 0 )
 		{
 			char c = *token.Begin();
 			const emoticonData_t *emoticon;
-			if ( c == '<' )
-			{
-				if ( span && !spanHasContent )
-				{
-					spanHasContent = true;
-					out.append( spanstr );
-				}
-				out.append( "&lt;" );
-			}
-			else if ( c == '>' )
-			{
-				if ( span && !spanHasContent )
-				{
-					spanHasContent = true;
-					out.append( spanstr );
-				}
-				out.append( "&gt;" );
-			}
-			else if ( c == '&' )
-			{
-				if ( span && !spanHasContent )
-				{
-					spanHasContent = true;
-					out.append( spanstr );
-				}
-				out.append( "&amp;" );
-			}
-			else if ( c == '\n' )
+
+			std::unordered_map<char, std::string> htmlEscapeStrings;
+			htmlEscapeStrings['<'] = "&lt;";
+			htmlEscapeStrings['>'] = "&gt;";
+			htmlEscapeStrings['&'] = "&amp;";
+			htmlEscapeStrings['"'] = "&quot;";
+			htmlEscapeStrings['\''] = "&apos;";
+			htmlEscapeStrings['\n'] = "<br />";
+
+			std::string htmlEscapedString = htmlEscapeStrings[ c ];
+
+			if ( htmlEscapedString != "" )
 			{
-				out.append( span && spanHasContent ? "</span><br />" : "<br />" );
-				span = false;
-				spanHasContent = false;
+				out.append( htmlEscapedString );
 			}
-			else if ( ( parseFlags & RP_EMOTICONS ) && ( emoticon = BG_EmoticonAt( token.Begin() ) ) )
+			else if ( emoticons && ( emoticon = BG_EmoticonAt( token.Begin() ) ) )
 			{
-				if ( span && !spanHasContent )
-				{
-					spanHasContent = true;
-					out.append( spanstr );
-				}
 				out.append( va( "<img class='emoticon' src='/%s' />", emoticon->imageFile.c_str() ) );
-				while ( *iter->Begin() != ']' )
+				while ( iter != parser.end() && *iter->Begin() != ']' )
 				{
 					++iter;
 				}
 			}
 			else
 			{
-				if ( span && !spanHasContent )
-				{
-					out.append( spanstr );
-					spanHasContent = true;
-				}
 				out.append( token.Begin(), token.Size() );
 			}
 		}
-		else if ( token.Type() == Color::Token::TokenType::COLOR )
+		else if ( quake && ( token.Type() == Color::Token::TokenType::COLOR ) )
```

Comment:
There is a behavior change here: we will add empty spans, ie

`^1^2Green` in the previous code led to:

```
<span style="color: green;">Green</span>
```

In the new code, this will be:

```
<span style="color: red;"></span><span style="color: green;">Green</span>
```

Should probably avoid doing this else people might abuse `^1^2^3^4^5^6hahah`

Reply:
Well, if we want to avoid that abuse, we would need convoluted code.

Reply:
Does the iterator support lookahead? (Or can we go backwards?) Can we do something like

```c++
++iter;
if (token.Type() == Color::Token::TokenType::COLOR)
{
  continue;
}
--iter;
char rgb[38];
...
```

Reply:
Right, we we may be able to look at the next token.